aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -0,0 +1,610 @@
+      <<<N * C, cuda_utils::kCUDABlockReduceNumThreads, 0, cuda_stream>>>(
+          HxW, dY_data, X_data, ds_data, db_data);
+  if (dX != nullptr) {
+    Tensor c1 = at::empty({N, C}, X.options().dtype(kAccType));
+    Tensor c2 = at::empty({N, G}, X.options().dtype(kAccType));
+    Tensor c3 = at::empty({N, G}, X.options().dtype(kAccType));
+    T_ACC* c1_data = c1.data_ptr<T_ACC>();
+    T_ACC* c2_data = c2.data_ptr<T_ACC>();
+    T_ACC* c3_data = c3.data_ptr<T_ACC>();
+    int64_t B = (N * C + kCUDANumThreads - 1) / kCUDANumThreads;
+    ComputeGradOutputCoeffientCUDAKernel<T>
+        <<<B, kCUDANumThreads, 0, cuda_stream>>>(
+            N, C, G, rstd_data, gamma_data, c1_data);
+    ComputeBackwardFusedParamsCUDAKernel<T>
+        <<<dim3(N, G),
+           cuda_utils::kCUDABlockReduceNumThreads,
+           0,
+           cuda_stream>>>(
+            C,
+            HxW,
+            G,
+            mean_data,
+            rstd_data,
+            gamma_data,
+            ds_data,
+            db_data,
+            c2_data,
+            c3_data);
+    if (HxW < kCUDANumThreads) {
+      B = (N * C * HxW + kCUDANumThreads - 1) / kCUDANumThreads;
+      GroupNormBackwardSimpleCUDAKernel<
+          T><<<B, kCUDANumThreads, 0, cuda_stream>>>(
+          N, C, HxW, G, dY_data, X_data, c1_data, c2_data, c3_data, dX_data);
+    } else {
+      GroupNormBackwardCUDAKernel<T>
+          <<<N * C, kCUDANumThreads, 0, cuda_stream>>>(
+              C, HxW, G, dY_data, X_data, c1_data, c2_data, c3_data, dX_data);
+    }
+  }
+  if (dgamma->defined() || dbeta->defined()) {
+    T* dgamma_data = dgamma->defined() ? dgamma->data_ptr<T>() : nullptr;
+    T* dbeta_data = dbeta->defined() ? dbeta->data_ptr<T>() : nullptr;
+    if (N < 512) {
+      // For small batch size, do colwise reduce directly.
+      const int64_t B = (C + kCUDANumThreads - 1) / kCUDANumThreads;
+      GammaBetaBackwardSimpleCUDAKernel<T>
+          <<<B, kCUDANumThreads, 0, cuda_stream>>>(
+              N,
+              C,
+              G,
+              mean_data,
+              rstd_data,
+              ds_data,
+              db_data,
+              dgamma_data,
+              dbeta_data);
+    } else {
+      const int64_t B = (C + kReduceTileSize - 1) / kReduceTileSize;
+      constexpr int kThreadX = kReduceTileSize;
+      constexpr int kThreadY = kReduceTileSize / 2;
+      GammaBetaBackwardCUDAKernel<T>
+          <<<B, dim3(kThreadX, kThreadY), 0, cuda_stream>>>(
+              N,
+              C,
+              G,
+              mean_data,
+              rstd_data,
+              ds_data,
+              db_data,
+              dgamma_data,
+              dbeta_data);
+    }
+  }
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void GroupNormBackwardKernelImpl(
+    const Tensor& dY,
+    const Tensor& X,
+    const Tensor& mean,
+    const Tensor& rstd,
+    const Tensor& gamma,
+    int64_t N,
+    int64_t C,
+    int64_t HxW,
+    int64_t group,
+    Tensor* dX,
```

Comment:
Why are you passing pointers here instead of non-const references? I think in the rest on the codebase we use references.

Reply:
Just to make it clear that pointers are for output arguments, and references are for inputs. This is something we used to do in Caffe2. We can consider to unify the styles in later PRs.